starboard/nplb/maximum_player_configuration_explorer.h

```diff
@@ -0,0 +1,133 @@
+
+#include "starboard/player.h"
+#include "starboard/shared/starboard/media/media_util.h"
+#include "starboard/testing/fake_graphics_context_provider.h"
+
+namespace starboard {
+namespace nplb {
+
+// The partial order relation here is: Given two vectors, a and b, a >= b if and
+// only if a[i] >= b[i] for all indices i.
+bool PosetGreaterThanOrEqualTo(const std::vector<int>& a,
+                               const std::vector<int>& b);
+
+// Given a vector v, test if it is contained within a max elements set s.
+// v is contained in a max elements set s if and only if there is a max element
+// m within s such that m >= v.
+bool PosetContainedByMaxElementsSet(
+    const std::set<std::vector<int>>& max_elements_set,
+    const std::vector<int>& v);
+
+// The scheduler is responsible for arranging the search process to identify the
+// maximum elements within a graded poset, which is based on a set of vectors
+// comprising integers. Specifically, the length of each vector is determined by
+// the number of |resource_types|, while the range of integers in each vector
+// spans from 0 to the maximum number of |max_instances_per_resource|, as [0,
+// max_instances]. The partial order <= between two vectors is defined such that
+// one vector is considered less than or equal to the other vector if all the
+// corresponding integers in the first vector are smaller than or equal to the
+// corresponding integers in the second vector. To identify the maximum
+// elements, a test function, denoted as |test_functor|, must be provided to
+// evaluate whether the current configuration satisfies the criterion for a
+// maximum element. The reason for applying the depth first search scheduler in
+// this context is that it is a faster method.
+std::set<std::vector<int>> ScheduleDFSOnGradedPoset(
+    int resource_types,
+    int max_instances_per_resource,
+    const std::function<bool(const std::vector<int>&)>& test_functor);
+
+// The |MaximumPlayerConfigurationExplorer| is employed to evaluate the maximum
+// feasible configurations for a group of players, with each player using a
+// distinct video/audio codec and output mode.
+// The search process is carried out using the |ScheduleDFSOnGradedPoset|, where
+// the |IsConfigsCreatable| function is provided as the test function.
+class MaximumPlayerConfigurationExplorer {
+ public:
+  typedef std::tuple<SbMediaVideoCodec,
+                     const shared::starboard::media::AudioStreamInfo,
+                     SbPlayerOutputMode>
+      PlayerConfig;
+
+  struct PlayerConfigSetComparator {
+    bool operator()(const PlayerConfig& lhs, const PlayerConfig& rhs) const {
+      const auto& audio_stream_info_lhs = std::get<1>(lhs);
+      const auto& audio_stream_info_rhs = std::get<1>(rhs);
+      return (std::get<0>(lhs) < std::get<0>(rhs)) ||  // SbMediaVideoCodec
+             (std::get<2>(lhs) < std::get<2>(rhs)) ||  // SbPlayerOutputMode
+             (audio_stream_info_lhs.codec <
```

Comment:
The comparison of all member variables within AudioStreamInfo has been incorporated. 

It is important to note that MaximumPlayerConfigurationExplorer is solely responsible for calculating the maximum player configuration, while the definition of the appropriate player configuration set remains the responsibility of the caller.

For example, if the caller think following two configs of the same audio codec = opus are different,
they should provide two.
``` 
1st  config
codec = opus 
number_of_channels = 6
samples_per_second = 44.1Kz
uint16_t bits_per_sample = 16

2nd  config
codec = opus 
number_of_channels = 2;
samples_per_second = 48kHz;
bits_per_sample = 16;
```

otherwise, they should provide the representative one, for example, 1st(since it support higher bitrate = 44.1K * 6 > 2 * 48K)
``` 
1st  config
codec = opus 
number_of_channels = 6
samples_per_second = 44.1Kz
uint16_t bits_per_sample = 16
```

Reply:
Usually when we compare a value with multiple dimensions, we'd have to do:
if (lhs.first < rhs.first) { return true; }
if (lhs.first > rhs.first) { return false; }
if (lhs.second < rhs.second) { return true; }
if (lhs.second > rhs.second) { return false; }
...

Otherwise, when we compare {1, 3} and {2, 2}, both "{1, 3} < {2, 2}" and "{2, 2} < {1, 3}" will be true, which isn't correct.

Reply:
Yes, I appreciate the suggestion.

In this case, we rely solely on the "==" operator since our data structure is a set. However, this approach can lead to inconsistencies if we modify the insertion order of the set.

Thank you for pointing out the correction. I appreciate it.